File: 0601-0700/0643-maximum-average-subarray-i.rs
struct Solution {}

impl Solution {
    pub fn find_max_average(nums: Vec<i32>, k: i32) -> f64 {
        let mut su = nums[..k as usize].iter().fold(0_f64, |acc, x| acc + *x as f64);
        nums[k as usize..].iter().zip(nums.iter()).fold(su, |acc, (x, old_x)| {
            su = su + (x - old_x) as f64;
            acc.max(su)
        }) / k as f64
    }
}

fn main() {
    let nums = vec![1,12,-5,-6,50,3];
    let k = 4;
    println!("{}", Solution::find_max_average(nums, k));
}

File: 0601-0700/0643.maximum-average-subarray-i.cpp
class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        if (nums.empty()) return 0;
        if (nums.size() < k)
            return accumulate(nums.begin(), nums.end(), 0.0) / nums.size();
        double su = accumulate(nums.begin(), nums.begin() + k, 0.0);
        double max_su = su;
        for (int i = k; i < nums.size(); ++i) {
            su = su + nums[i] - nums[i - k];
            max_su = max(max_su, su);
        }
        return max_su / k;
    }
};
